crates/kubelet/src/volume/persistentvolumeclaim.rs

```diff
@@ -0,0 +1,319 @@
+    let mut stage_unstage_volume = false;
+    let response = csi_client
+        .node_get_capabilities(NodeGetCapabilitiesRequest {})
+        .await?;
+    for capability in &response.get_ref().capabilities {
+        if let Some(typ) = &capability.r#type {
+            let _typ_stage_unstage_volume = rpc::Type::StageUnstageVolume as i32;
+            match typ {
+                CapabilityType::Rpc(Rpc {
+                    r#type: _typ_stage_unstage_volume,
+                }) => {
+                    stage_unstage_volume = true;
+                }
+            }
+        }
+    }
+
+    // we keep this around even if the driver does not support STAGE_UNSTAGE_VOLUME. unmount() still needs it.
+    tokio::fs::create_dir_all(path).await?;
+
+    // https://github.com/kubernetes/kubernetes/blob/734889ed822d1a60c6dd61ccd8f1ed0e8ab31ea5/pkg/volume/csi/csi_attacher.go#L295-L298
+    let csi = pv.spec.unwrap().csi.unwrap();
+    let data = serde_json::json!({
+        "volumeHandle": csi.volume_handle,
+        "driverName": csi.driver,
+    });
+
+    // save volume data to path
+    // https://github.com/kubernetes/kubernetes/blob/734889ed822d1a60c6dd61ccd8f1ed0e8ab31ea5/pkg/volume/csi/csi_attacher.go#L294-L305
+    let mut vol_data_filepath = path.clone();
+    vol_data_filepath.push("../".to_owned() + VOLUME_DATA_FILE_NAME);
+    tokio::fs::write(&vol_data_filepath, serde_json::to_string(&data)?).await?;
+
+    // https://github.com/kflansburg/k8s-csi/blob/85a1e81757332724f5e23b452fdb2da92f0c1923/proto/csi/v1.3.0/api.proto#L1189
+    let staging_path = path.parent().unwrap().join("_staging");
+    let mut req = NodePublishVolumeRequest {
+        volume_id: csi.volume_handle.clone(),
+        target_path: path.to_string_lossy().to_string(),
+        staging_target_path: "".to_owned(),
+        volume_capability: Some(VolumeCapability {
+            // TODO: determine the correct access mode and mount flags from the volume
+            // https://github.com/kubernetes/kubernetes/blob/734889ed822d1a60c6dd61ccd8f1ed0e8ab31ea5/pkg/volume/csi/csi_attacher.go#L325-L333
+            access_mode: Some(CSIAccessMode {
+                mode: CSIMode::SingleNodeWriter as i32,
+            }),
+            access_type: Some(CSIAccessType::Mount(CSIMountVolume {
+                fs_type: csi.fs_type.clone().unwrap_or("".to_owned()),
+                mount_flags: Default::default(),
+            })),
+        }),
+        // hardcode to read/write for now
+        // TODO: determine the correct access mode and mount flags from the volume
+        // https://github.com/kubernetes/kubernetes/blob/734889ed822d1a60c6dd61ccd8f1ed0e8ab31ea5/pkg/volume/csi/csi_attacher.go#L325-L333
+        readonly: false,
+        secrets: Default::default(),
+        publish_context: Default::default(),
+        volume_context: Default::default(),
+    };
+
+    if stage_unstage_volume {
+        debug!("stage/unstage supported");
```

Comment:
Can you add more context here?

Reply:
garden variety debug statement; will remove